### ISSUE TITLE
Mod Loader

### ISSUE THREAD

**Issue by SimonMeskens** [BTW-Community/BTW-Public#24]
There are two parts to this:

* Being able to bring class files into the classpath, so you don't need to register a mod with base class edit anymore
* Being able to replace base class files with external files, so that mods can just go in a /mods/ folder

Java's way of handling the loading of classes is aptly called [ClassLoader](https://docs.oracle.com/javase/6/docs/api/java/lang/ClassLoader.html). Making a custom one isn't hard, there's facilities to pull in an entire jar (or zip?) at once, etc, but it also ships with [URLClassLoader](https://docs.oracle.com/javase/6/docs/api/java/net/URLClassLoader.html).

The javadocs give the following example for implementing your own:
```java
class NetworkClassLoader extends ClassLoader {
     String host;
     int port;

     public Class findClass(String name) {
         byte[] b = loadClassData(name);
         return defineClass(name, b, 0, b.length);
     }

     private byte[] loadClassData(String name) {
         // load the class data from the connection
          . . .
     }
 }
```

To make the class loader replace a class at runtime is also not hard, you just have to [overload the `loadClass` method](https://stackoverflow.com/a/14317506), but here we run into one subtle issue:

Every class is handled by exactly one class loader, by default that would be the "System class loader". To get control over the loading of the default Minecraft classes, we need to [hook into the Main function](https://stackoverflow.com/a/148762), before any classes are loaded, set up a proxy class loader, then instantiate the real Main function through this proxy. Once the codebase is being loaded through a ProxyClassLoader we control, we can have that ProxyClassLoader load the vanilla codebase in one ClassLoader and each separate mod jar in its own [JarClassLoader](https://docs.oracle.com/javase/tutorial/deployment/jar/apiindex.html). When a JarClassLoader is done loading, we check if it has base classes, register those in a map, and then when a class is requested, the Proxy looks at the map first, before delegating to vanilla. Doing it this way even gives you a way to just give a warning (or error) on load that two different mods are both trying to replace the same base class. In some cases that might be desirable however (patches), so there could also be a simple system able to resolve conflicts.

To summarize:
* Replace Main with a new method that starts a ProxClassLoader
* For each mod jar, load classes in a separate JarClassLoader
* Look through the JarClassLoaders and find all the classes that are base class edits (aka, their full name already exists in VanillaClassLoader), add them to a map in ProxyClassLoader
* Start VanillaClassloader which has access to all of vanilla's classes
* Start the Main through ProxyClassLoader, when requests for classes come in, delegate to the correct class in the correct Class Loader

Result: mods can be in their own jar (or zip) in the /mods/ folder and can contain edits of base classes that get loaded in

https://github.com/BTW-Community/BTW-source/issues/64

**Comment by SimonMeskens** [comment]
We probably also want a system with Proxy classes, so that mods can override just a single method.

**Comment by ammoore00** [comment]
Closing this as migration to fabric handles this.
